core/state/pruner/pruner.go

```diff
@@ -0,0 +1,517 @@
+	if err != nil {
+		return nil, err
+	}
+	return &Pruner{
+		db:            db,
+		stateBloom:    stateBloom,
+		datadir:       datadir,
+		trieCachePath: trieCachePath,
+		headHeader:    headHeader,
+		snaptree:      snaptree,
+	}, nil
+}
+
+func prune(maindb ethdb.Database, stateBloom *stateBloom, start time.Time) error {
+	// Extract all node refs belong to the genesis. We have to keep the
+	// genesis all the time.
+	genesisHashset, err := extractGenesis(maindb)
+	if err != nil {
+		return err
+	}
+	// Delete all stale trie nodes in the disk. With the help of state bloom
+	// the trie nodes(and codes) belong to the active state will be filtered
+	// out. A very small part of stale tries will also be filtered because of
+	// the false-positive rate of bloom filter. But the assumption is held here
+	// that the false-positive is low enough(~0.05%). The probablity of the
+	// dangling node is the state root is super low. So the dangling nodes in
+	// theory will never ever be visited again.
+	var (
+		count  int
+		size   common.StorageSize
+		pstart = time.Now()
+		logged = time.Now()
+		batch  = maindb.NewBatch()
+		iter   = maindb.NewIterator(nil, nil)
+
+		rangestart, rangelimit []byte
+	)
+	for iter.Next() {
+		key := iter.Key()
+
+		// All state entries don't belong to specific state and genesis are deleted here
+		// - trie node
+		// - legacy contract code
+		// - new-scheme contract code
+		isCode, codeKey := rawdb.IsCodeKey(key)
+		if len(key) == common.HashLength || isCode {
+			checkKey := key
+			if isCode {
+				checkKey = codeKey
+			}
+			// Filter out the state belongs to the genesis
+			if _, ok := genesisHashset[common.BytesToHash(checkKey)]; ok {
+				continue
+			}
+			// Filter out the state belongs the pruning target
+			if ok, err := stateBloom.Contain(checkKey); err != nil {
+				return err
+			} else if ok {
+				continue
+			}
+			// Filter out the state belongs to the "blacklist". Usually
+			// the root of the "useless" states are contained here.
+			size += common.StorageSize(len(key) + len(iter.Value()))
+			batch.Delete(key)
+
+			if batch.ValueSize() >= ethdb.IdealBatchSize {
+				batch.Write()
+				batch.Reset()
+			}
+			count += 1
+			if count%1000 == 0 && time.Since(logged) > 8*time.Second {
+				log.Info("Pruning state data", "count", count, "size", size, "elapsed", common.PrettyDuration(time.Since(pstart)))
+				logged = time.Now()
+			}
+			if rangestart == nil || bytes.Compare(rangestart, key) > 0 {
+				if rangestart == nil {
+					rangestart = make([]byte, common.HashLength)
+				}
+				copy(rangestart, key)
+			}
+			if rangelimit == nil || bytes.Compare(rangelimit, key) < 0 {
+				if rangelimit == nil {
+					rangelimit = make([]byte, common.HashLength)
+				}
+				copy(rangelimit, key)
+			}
+		}
+	}
+	if batch.ValueSize() > 0 {
+		batch.Write()
+		batch.Reset()
+	}
+	iter.Release() // Please release the iterator here, otherwise will block the compactor
```

Comment:
Wondering if we should recreate the iterator after every batch commit perhaps? Otherwise all those batches will just pile up in some LevelDB limbo and won'g get processed until we reach this point. Maybe I'm wrong here, so feel free to object.

Reply:
yes, make sense